fdbclient/include/fdbclient/KeyBackedTypes.actor.h

```diff
@@ -0,0 +1,1093 @@
+	template <class Transaction>
+	typename std::enable_if<!is_transaction_creator<Transaction>, Future<T>>::type
+	getD(Transaction tr, Snapshot snapshot = Snapshot::False, T defaultValue = T()) const {
+		return map(get(tr, snapshot), [=](Optional<T> val) -> T { return val.present() ? val.get() : defaultValue; });
+	}
+
+	// Get property's value or throw error if it doesn't exist
+	template <class Transaction>
+	typename std::enable_if<!is_transaction_creator<Transaction>, Future<T>>::type
+	getOrThrow(Transaction tr, Snapshot snapshot = Snapshot::False, Error err = key_not_found()) const {
+		return map(get(tr, snapshot), [=](Optional<T> val) -> T {
+			if (!val.present()) {
+				throw err;
+			}
+
+			return val.get();
+		});
+	}
+
+	template <class DB>
+	typename std::enable_if<is_transaction_creator<DB>, Future<Optional<T>>>::type get(
+	    Reference<DB> db,
+	    Snapshot snapshot = Snapshot::False) const {
+		return runTransaction(db, [=, self = *this](Reference<typename DB::TransactionT> tr) {
+			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+
+			return self.get(tr, snapshot);
+		});
+	}
+
+	template <class DB>
+	typename std::enable_if<is_transaction_creator<DB>, Future<T>>::type getD(Reference<DB> db,
+	                                                                          Snapshot snapshot = Snapshot::False,
+	                                                                          T defaultValue = T()) const {
+		return runTransaction(db, [=, self = *this](Reference<typename DB::TransactionT> tr) {
+			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+
+			return self.getD(tr, snapshot, defaultValue);
+		});
+	}
+
+	template <class DB>
+	typename std::enable_if<is_transaction_creator<DB>, Future<T>>::type getOrThrow(Reference<DB> db,
+	                                                                                Snapshot snapshot = Snapshot::False,
+	                                                                                Error err = key_not_found()) const {
+		return runTransaction(db, [=, self = *this](Reference<typename DB::TransactionT> tr) {
+			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+
+			return self.getOrThrow(tr, snapshot, err);
+		});
+	}
+
+	template <class Transaction>
+	typename std::enable_if<!is_transaction_creator<Transaction>, void>::type set(Transaction tr, T const& val) {
+		tr->set(key, packValue(val));
+		if (trigger.present()) {
+			trigger.get().update(tr);
+		}
+	}
+
+	template <class DB>
+	typename std::enable_if<is_transaction_creator<DB>, Future<Void>>::type set(Reference<DB> db, T const& val) {
+		return runTransaction(db, [=, self = *this](Reference<typename DB::TransactionT> tr) {
+			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
```

Comment:
does this need PRIORITY_SYSTEM_IMMEDIATE?

Reply:
@ajbeamon Since these classes are generally used in the system space, should we set `PRIORITY_SYSTEM_IMMEDIATE` for all of the `DB` interfaces?  At least some of the use cases would want to work when the cluster is having problems.

Reply:
Hmm, I'm not sure whether we do want that or not for some of the tenant stuff. If we used this functionality that a client could trigger it, then it could mean that they could do arbitrary amounts of tenant metadata reads (or similarly metacluster metadata reads) without us being able to throttle them. Maybe we want that for all the operations, or possibly some but not all?

It's also the case that in the metacluster, we are using these types outside of the system key-space. I'm not sure that really has much bearing on the answer though.

Reply:
I really don't like the verbosity of having to write an explicit retry loop just to set an option, but I also think it's unsafe to set these options for all use cases.  `ACCESS_SYSTEM_KEYS` or `READ_SYSTEM_KEYS` are fairly easy to justify for these classes, but `PRIORITY_SYSTEM_IMMEDIATE` and `LOCK_AWARE` are probably going too far.

I'm going to try to make a templated `Database` wrapper for system transactions that provides a syntactically short way to specify these options and works with DB types and supplies what is needed by `KeyBackedTypes` and `RunTransaction` methods.
..._time passes_...
Okay here's what I've come up with:

A function `SystemDB(db, write, lock, immediate)` will return the DB wrapper.  The flags are optional and default to false, leaving only `READ_SYSTEM_KEYS` set in the transactions produced.  I also created `SystemDBLockWriteNow()` which in addition to being a really fun pun will run SystemDB with with all flags true.

For now, to not break anything, I will leave all existing `KeyBacked..` methods and their post-refactor equivalents alone because removing the lock/system options on these would certainly break things.  

For the new `KeyBackedRangeMap` class I will not set any transaction options in its DB methods and rely on the new wrapper types to be invoked at the call sites instead.

Reply:
This solution almost works, but unfortunately the loops in the RunTransaction methods only create a transaction once and reuse it in the retry loop, which makes sense as the backoff timer state needs to be maintained but it turns out that `onError()` resets the transaction options.  @ajbeamon Is this an essential behavior?  While obviously a retry may not call the same methods / set the same transaction options, it does not seem out of line to me for `.onError()` to keep the transaction options, or at least these particular system related flags.

Reply:
I've fixed that problem by creating a new trait similar to `is_transaction_creator<>` which is `can_set_transaction_options<>`, which is `false` by default but `true` for the DB wrapper type which knows which options to set.  

I've modified both `runTransaction()` templates to do a `constexpr` check that the db reference given can set transaction options, and if so it calls `db->setOptions(tr)` at the top of the retry loop.  The manual transaction loops in KeyBackedTypes also call `setOptions()` if possible.

I also had to add  `if constexpr (..` support to the ActorCompiler for all this to work.

Reply:
Renamed `SystemDBLockWriteNow` to `SystemDBWriteLockedNow` which I think is a bit more clear.